File: USACO/Silver/Python/USACO 2023 Jan Silver Following Directions.py
import sys

input = sys.stdin.readline

n = int(input())
grid = [[None]*(n+1) for _ in range(n+1)]

for i in range(n):
    directions, cost = map(str, input().split())
    grid[i][:n] = [c for c in directions]
    grid[i][n] = int(cost)
grid[n] = [int(i) for i in input().split()]

dp = [[0]*(n+1) for _ in range(n+1)]

def calc():
    ret = 0
    for i in range(n):
        ret += dp[i][n] * grid[i][n] + dp[n][i] * grid[n][i]
    return ret

def update(i, j, v):
    while True:
        dp[i][j] += v
        if i == n or j == n:
            return
        if grid[i][j] == 'R':
            j += 1
        else:
            i += 1

for i in range(n+1):
    for j in range(n+1):
        if i != n and j != n:
            dp[i][j] = 1
        if i != 0 and j != n and grid[i-1][j] == 'D':
            dp[i][j] += dp[i-1][j]
        if i != n and j != 0 and grid[i][j-1] == 'R':
            dp[i][j] += dp[i][j-1]

print(calc())

q = int(input())
for _ in range(q):
    i, j = map(int, input().split())
    i -= 1
    j -= 1
    cur = dp[i][j]
    update(i, j, -cur)
    if grid[i][j] == 'R':
        grid[i][j] = 'D'
    else:
        grid[i][j] = 'R'
    update(i, j, cur)
    print(calc())

File: USACO/Silver/C++/USACO 2023 Jan Silver Following Directions.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define MOD(a, b) ((((a) % (b)) + (b)) % (b))

bool cbs(pair<int, int> a, pair<int, int> b) {
    return a.second < b.second;
}

signed main() {
    //freopen("input.in", "r", stdin);
    //freopen("output.out", "w", stdout);

    cin.tie(NULL) -> sync_with_stdio(false);
    
    // wtf is this time limit
    // this is kinda like dsu except we need to break connections?
    // since theres only r and d surely that means we can somehow prefix sum this
    // for fullsolve this has to be at most O(n^2q)
    // ok i think i got it
    // we keep track of how many cells point toward that one cell
    // and then for all the edge cells we can just multiply expected
    // O(n^2) update should be acceptable, we just dfs and update both sides
    // implementation is pain wtf
    int n; cin >> n;
    vector<vector<char>> grid(n, vector<char>(n));
    vector<int> rval(n), dval(n);
    for(int i=0; i<n; ++i){
        for(int j=0; j<n; ++j) cin >> grid[i][j];
        cin >> rval[i];
    }
    for(int i=0; i<n; ++i) cin >> dval[i];
    
    vector<vector<int>> prop(n, vector<int>(n, 1));
    for(int i=0; i<n; ++i){
        for(int j=0; j<n; ++j){
            if (grid[i][j] == 'R'){
                if (j+1 < n) prop[i][j+1] += prop[i][j];
            }else{
                if(i+1 < n) prop[i+1][j] += prop[i][j];
            }
        }
    }
    ll ans=0;
    for(int i=0; i<n; ++i) if(grid[i][n-1] == 'R') ans += prop[i][n-1]*rval[i];
    for(int i=0; i<n; ++i) if(grid[n-1][i] == 'D') ans += prop[n-1][i]*dval[i];
    cout << ans << '\n';
    
    int q; cin >> q;
    while(q--){
        int ti, tj;
        cin >> ti >> tj;
        --ti; --tj;
        grid[ti][tj] = (grid[ti][tj] == 'R') ? 'D':'R';
        int pval = prop[ti][tj];
        
        int pi = ti+(grid[ti][tj] == 'D');
        int pj = tj+(grid[ti][tj] == 'R');
        int ni = ti+(grid[ti][tj] == 'R');
        int nj = tj+(grid[ti][tj] == 'D');
        while(pi<n && pj<n) {
            prop[pi][pj] += pval;
            if (grid[pi][pj] == 'R') ++pj;
            else ++pi;
        }
        while(ni<n && nj<n) {
            prop[ni][nj] -= pval;
            if (grid[ni][nj] == 'R') ++nj;
            else ++ni;
        }
        ans = 0;
        for(int i=0; i<n; ++i) if(grid[i][n-1] == 'R') ans += prop[i][n-1]*rval[i];
        for(int i=0; i<n; ++i) if(grid[n-1][i] == 'D') ans += prop[n-1][i]*dval[i];
        cout << ans << '\n';
    }
}
